File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        llist = []
        def dfs(root):
            if root is None:
                return
            llist.append(root.val)
            dfs(root.left)
            dfs(root.right)
        
        dfs(root)
        heap = []
        for num in llist:
            heapq.heappush(heap, num)

        res = 0
        
        for i in range(k):
            res = heapq.heappop(heap)
        return res

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxPathSum(TreeNode root) {
        int[] res = new int[]{Integer.MIN_VALUE};
        if(root == null) return res[0];

        dfs(root, res);
        return res[0];
    }

    private int dfs(TreeNode root, int[] res) {
        if(root == null) return Integer.MIN_VALUE;

        int leftPathSum = dfs(root.left, res);
        if(leftPathSum < 0) leftPathSum = 0;

        int rightPathSum = dfs(root.right, res);
        if(rightPathSum < 0) rightPathSum = 0;

        res[0] = Math.max(res[0], root.val + leftPathSum + rightPathSum);

        return root.val + Math.max(leftPathSum, rightPathSum);
    }
}
